src/bigtiffimage.cpp

```diff
@@ -226,13 +226,24 @@ namespace Exiv2
                                                   : is8ByteType(type)  ? 8
                                                   : 1;
 
-                			// #55 memory allocation crash test/data/POC8
-                			long long allocate = (long long) size*count + pad;
-                			if ( allocate > (long long) io.size() ) {
-                    			throw Error(57);
-                			}
+                            // #55 and #56 memory allocation crash test/data/POC8
 
-                            DataBuf buf((long)allocate);
+                            // size * count > std::numeric_limits<uint64_t>::max()
+                            // =>
+                            // size > std::numeric_limits<uint64_t>::max() / count
+                            if (size > std::numeric_limits<uint64_t>::max() / count)
+                                throw Error(57);             // we got number bigger than 2^64
+                                                             // more than we can handle
+
+                            if (size * count > std::numeric_limits<uint64_t>::max() - pad)
```

Comment:
It is not the same, imagine the example `size = 1` and `count = 0xFFFF FFFF`. Then `size * count` will not overflow, but `size * count + pad` will overflow for `pad = 1` and result in `allocate = 0`. Your proposed check `allocate > std::numeric_limits<uint64_t>::max()` is never true, since allocate is also an uint64_t. The checks that @Kicer86 wrote work around this nicely.

Reply:
Thanks again. I never tried to fix this kind of issues. It is really interesting to learn this kind of stuff :)